### ISSUE TITLE
grammar not capturing struct

### ISSUE THREAD

**Issue by djacobow** [mohawk2/Inline-Struct#6]
Hi again! Thanks for taking a looking at this!

The struct fparams_t below is really the one I want Inline::Struct to capture correctly.

You can also see this code in https://github.com/djacobow/mandeldemo/tree/master/native. It is the same except it uses types from stdint.h that Perl doesn't understand without extra typemaps.

And here is the piece I've been working on, if you're interested: https://toolsofourtools.org/archives/769

fractizer.h:

```
#ifndef __FRACTIZER_H
#define __FRACTIZER_H

#include <math.h>
#include <stdint.h>

#define INLINEIT static inline

#ifdef __cplusplus
extern "C" {
#endif

typedef struct cp_t {
 double r;
 double i;
} cp_t;

typedef void (*znp1_calc_t)(cp_t *z, cp_t *c);

typedef struct fparams_t {
    unsigned short max_iters;
    double   escape_val;
    double   x_min;
    double   x_max;
    double   y_min;
    double   y_max;
    unsigned short x_pels;
    unsigned short y_pels;
    unsigned short x_tile;
    unsigned short y_tile;
    unsigned char  type;
    unsigned char  do_julia;
    double   jx;
    double   jy;
    znp1_calc_t algo;
} fparams_t;
void showParams(fparams_t *p);
void set_default_params(fparams_t *p);
void generate_fractal(fparams_t *pparams, unsigned short *rbuf);

#ifdef __cplusplus
}
#endif

#endif
```

Just for completeness I am including the rest of the code to build a quick-n-dirty demo under Inline::C

fractizer.c:

```
#include "fractizer.h"
#include <stdio.h>

INLINEIT double cmag(cp_t *z) {
    return sqrt(z->r * z->r + z->i * z->i);
}
INLINEIT cp_t csquare(cp_t *a) {
    cp_t r;
    r.r = a->r * a->r - a->i * a->i;
    r.i = 2 * a->r * a->i;
    return r;
}

INLINEIT cp_t ccube(cp_t *z) {
    cp_t r;
    r.r = z->r * z->r * z->r - 3 * z->r * z->i * z->i;
    r.i = 3 * z->r * z->r * z->i - z->i * z->i * z->i;
    return r;
};

INLINEIT cp_t cadd(cp_t *a, cp_t *b) {
    cp_t r;
    r.r = a->r + b->r;
    r.i = a->i + b->i;
    return r;
};

INLINEIT cp_t csmul(cp_t *a, double m) {
    cp_t r;
    r.r = a->r * m;
    r.i = a->i * m;
    return r;
}

INLINEIT unsigned char calc_pixel_generic(fparams_t *pparams, double x, double y, unsigned short *piters, double *pval) {
    cp_t z, c;
    if (pparams->do_julia) {
        z.r = x; z.i = y; c.r = pparams->jx; c.i = pparams->jy;
    } else {
        z.r = 0; z.i = 0; c.r = x; c.i = y;
    };
    unsigned short iter = 0;
    double val = 0;
    while ((iter < pparams->max_iters) && (val < pparams->escape_val)) {
        pparams->algo(&z, &c);
        val = cmag(&z);
        iter++;
    }
    *pval = val;
    *piters = iter;
    return iter < pparams->max_iters;
};

INLINEIT void znp1_mb(cp_t *z, cp_t *c) {
    cp_t nz;
    double n_z_r = z->r * z->r - z->i * z->i + c->r;
    nz.i = 2.0 * z->r * z->i + c->i;
    nz.r = n_z_r;
    *z = nz;
}

INLINEIT void znp1_cos(cp_t *z, cp_t *c) {
    cp_t nz;
    nz.r = cos(z->r) * cosh(z->i) + c->r;
    nz.i = -sin(z->r) * sinh(z->i) + c->i;
    *z = nz;
}

INLINEIT void znp1_mbtc(cp_t *z, cp_t *c) {
    z->i = -z->i;
    cp_t nz;
    nz.r = z->r * z->r - z->i * z->i + c->r;
    nz.i = 2.0 * z->r * z->i + c->i;
    *z = nz;;
}

INLINEIT void znp1_mbbs(cp_t *z, cp_t *c) {
    z->r = fabs(z->r);
    z->i = fabs(z->i);
    cp_t nz;
    nz.r = z->r * z->r - z->i * z->i + c->r;
    nz.i = 2.0 * z->r * z->i + c->i;
    *z = nz;;
}

INLINEIT void znp1_mb3(cp_t *z, cp_t *c) {
    cp_t nz;
    nz = ccube(z);
    nz.r += c->r;
    nz.i += c->i;
    *z = nz;;
}

INLINEIT void znp1_nf3(cp_t *z, cp_t *c) {
    cp_t nz = ccube(z);
    nz.r += c->r - 1.0;
    nz.i += c->i;
    *z = nz;
}

INLINEIT void znp1_nf3m2z(cp_t *z, cp_t *c) {
    cp_t a = ccube(z);
    cp_t b = csmul(z,-2);
    cp_t nz = cadd(&a,&b);
    nz.r += c->r;
    nz.i += c->i;
    *z = nz;
};

INLINEIT void znp1_nf8p15z4(cp_t *z, cp_t *c) {
    cp_t z2 = csquare(z);
    cp_t z4 = csquare(&z2);
    cp_t z8 = csquare(&z4);
    cp_t z14_15 = csmul(&z4,15);
    cp_t nz = cadd(&z8,&z14_15);
    nz.r += c->r - 0;
    nz.i += c->i;
    *z = nz;
}

void generate_fractal(fparams_t *pparams, unsigned short *rbuf) {
    double x_step = (pparams->x_max - pparams->x_min) / (double)pparams->x_pels;
    double y_step = (pparams->y_max - pparams->y_min) / (double)pparams->y_pels;

    pparams->algo = znp1_mb;

    switch (pparams->type) {
        case 1 : pparams->algo = znp1_mb3; break;
        case 2 : pparams->algo = znp1_mbbs; break;
        case 3 : pparams->algo = znp1_mbtc; break;
        case 4 : pparams->algo = znp1_cos; break;
        case 5 : pparams->algo = znp1_nf3; break;
        case 6 : pparams->algo = znp1_nf3m2z; break;
        case 7 : pparams->algo = znp1_nf8p15z4; break;
        default: pparams->algo = znp1_mb; break;
    }

    for (unsigned short j=0; j<pparams->y_pels; j++) {
        double y = pparams->y_min + j * y_step;
        for (unsigned short i=0; i<pparams->x_pels; i++) {
            double x = pparams->x_min + i * x_step;
            unsigned short iters = 0;
            double val = 0;
            calc_pixel_generic(pparams, x, y, &iters, &val);
            rbuf[i + j * pparams->x_pels] = iters;
        }
    };
};

void set_default_params(fparams_t *p) {
    p->max_iters = 200;
    p->escape_val = 4;
    p->x_min = -2;
    p->x_max = 2;
    p->y_min = -2;
    p->y_max = 2;
    p->x_pels = 1024;
    p->y_pels = 1024;
    p->x_tile = 0;
    p->y_tile = 0;
    p->type = 0;
    p->do_julia = 0;
    p->jx = 0;
    p->jy = 0;
};

void showParams(fparams_t *p) {
    printf("max_iters   : %d\n",p->max_iters);
    printf("escape_Val  : %f\n",p->escape_val);
    printf("x_range     : [%f : %f ]\n",p->x_min,p->x_max);
    printf("y_range     : [%f : %f ]\n",p->y_min,p->y_max);
    printf("size        : (%d : %d )\n",p->x_pels,p->y_pels);
    printf("tiles       : (%d : %d )\n",p->x_tile,p->y_tile);
    printf("type        : %d\n",p->type);
    printf("do_julia    : %d\n",p->do_julia);
    printf("jparams     : (%f : %f )\n",p->jx,p->jy);
};

```

perl_typemaps:
```
fparams_t *    T_PTR
cp_t *         T_PTR
uint16_t *     T_PTR

uint32_t       T_UV
uint16_t       T_UV
int32_t        T_IV
int16_t        T_IV
uint8_t        T_U_CHAR
int8_t         T_CHAR
```

perl_ex.pl:
```
#!/usr/bin/perl -w
  
use strict;
use warnings qw(all);
use Data::Dumper;
use Inline C =>
    Config =>
        INC => '.',
        TYPEMAPS => 'perl_typemaps',
        ENABLE => "AUTOWRAP";

use Inline "C";
Inline->init;


my $width = 120;
my $height = 60;


sub showResult {
    my $output = shift;
    my $olines = [];
    for (my $j=0;$j<$height;$j++) {
        my $line = '';
        for (my $i=0;$i<$width;$i++) {
            my $v = $output->[$j*$width+$i];
            my $s = $v >= 200 ? '*' : ' ';
            $line .= $s;
        };
        push(@$olines,$line);
    };
    print(join("\n",@$olines));
    print("\n");
};

my $params = new_fparams();
fr_set_x_pels($params,$width);
fr_set_y_pels($params,$height);
fr_show($params);
my $output = fr_calc($params);
showResult($output);

1;

__DATA__
__C__

#include "fractizer.h"
#include "fractizer.c"

fparams_t *new_fparams() {
    fparams_t *p = malloc(sizeof(fparams_t));
    if (p) set_default_params(p);
    return p;
}

void free_fparams(fparams_t *p) {
    if (p) free(p);
}

void fr_set_max_iters(fparams_t *p, uint16_t i) { p->max_iters = i; };
void fr_set_escape_val(fparams_t *p, double d)  { p->escape_val = d; };
void fr_set_x_min(fparams_t *p, double d)       { p->x_min = d; };
void fr_set_x_max(fparams_t *p, double d)       { p->x_max = d; };
void fr_set_y_min(fparams_t *p, double d)       { p->y_min = d; };
void fr_set_y_max(fparams_t *p, double d)       { p->y_max = d; };
void fr_set_x_pels(fparams_t *p, uint16_t i)    { p->x_pels = i; };
void fr_set_y_pels(fparams_t *p, uint16_t i)    { p->y_pels = i; };
void fr_set_x_tile(fparams_t *p, uint16_t i)    { p->x_tile = i; };
void fr_set_y_tile(fparams_t *p, uint16_t i)    { p->y_tile = i; };
void fr_set_type(fparams_t *p, uint8_t i)       { p->type = i; };
void fr_set_do_julia(fparams_t *p, uint8_t i)   { p->do_julia = i; };
void fr_set_jx(fparams_t *p, double d)          { p->jx = d; };
void fr_set_jy(fparams_t *p, double d)          { p->jy = d; };

void fr_show(fparams_t *p) {
    showParams(p);
};

SV *fr_calc(fparams_t *p) {
    size_t len = p->x_pels * p->y_pels;
    uint16_t *buf = malloc(sizeof(uint16_t) * len);
    generate_fractal(p,buf);

    AV* array = newAV();
    for (size_t i=0; i<len; i++) {
        av_push(array, newSVuv(buf[i]));
    };
    free(buf);
    return newRV_noinc((SV*)array);
}
```

**Comment by mohawk2** [comment]
Looks like the reason it's not capturing `fparams_t` is because of the function-pointer member.

**Comment by mohawk2** [comment]
Fixed that. Released as 0.27. Let me know (feel free to reopen) if any problems.
